lib/class-wp-rest-url-details-controller.php

```diff
@@ -0,0 +1,291 @@
+						),
+					),
+					'permission_callback' => array( $this, 'permissions_check' ),
+					'schema'              => array( $this, 'get_public_item_schema' ),
+				),
+			)
+		);
+	}
+
+	/**
+	 * Get the schema for the endpoint.
+	 *
+	 * @return array the schema.
+	 */
+	public function get_item_schema() {
+
+		if ( $this->schema ) {
+			return $this->add_additional_fields_schema( $this->schema );
+		}
+
+		$schema = array(
+			'$schema'    => 'http://json-schema.org/draft-04/schema#',
+			'title'      => 'url-details',
+			'type'       => 'object',
+			'properties' => array(
+				'title' => array(
+					'description' => __( 'The contents of the <title> tag from the URL.', 'gutenberg' ),
+					'type'        => 'string',
+					'context'     => array( 'view', 'edit', 'embed' ),
+					'readonly'    => true,
+				),
+			),
+		);
+
+		$this->schema = $schema;
+
+		return $this->add_additional_fields_schema( $this->schema );
+	}
+
+	/**
+	 * Retrieves the contents of the <title> tag from the HTML
+	 * response.
+	 *
+	 * @param WP_REST_REQUEST $request Full details about the request.
+	 * @return String|WP_Error The title text or an error.
+	 */
+	public function parse_url_details( $request ) {
+
+		$url = untrailingslashit( $request->get_param( 'url' ) );
+
+		if ( empty( $url ) ) {
+			return new WP_Error( 'rest_invalid_url', __( 'Invalid URL', 'gutenberg' ), array( 'status' => 404 ) );
+		}
+
+		// Transient per URL.
+		$cache_key = $this->build_cache_key_for_url( $url );
+
+		// Attempt to retrieve cached response.
+		$cached_response = $this->get_cache( $cache_key );
+
+		if ( ! empty( $cached_response ) ) {
+			$from_cache = true;
+			$response = rest_ensure_response( json_decode( $cached_response, true ) );
+
+			/**
+			 * Filters the URL data for the response.
+			 *
+			 * @param WP_REST_Response $response The response object.
+			 * @param string           $url      The requested URL.
+			 * @param bool             $from_cache Whether the response is from the cache.
+			 * @param WP_REST_Request  $request  Request object.
+			 */
+			return apply_filters( 'rest_url_details_prepare_response', $response, $url, $from_cache, $request );
+		}
+
+		// If we're this far the response is not from the cache.
+		$from_cache = false;
+
+		$response_body = $this->get_remote_url( $url );
+
+		// Exit if we don't have a valid body or it's empty.
+		if ( is_wp_error( $response_body ) || empty( $response_body ) ) {
+			return $response_body;
+		}
+
+		/**
+		 * Filters the data returned for a given URL.
+		 *
+		 * Can be used to modify the data retrieved for a given URL.
+		 * For example, you may wish to retrieve additional information
+		 * from the response body at a given URL over and above the `<title>`.
+		 *
+		 * @param array $data The default data retrieved for the given URL.
+		 * @param array $response_body The response body (HTML) for the given URL.
+		 * @param string $url The attempted URL.
+		 */
+		$data = apply_filters(
```

Comment:
This filter shouldn't exist anymore. The general response filter can be used.

Reply:
I think it might still be required as this is 

1. Filtering the data that is going into cache.
2. Filtering directly on the response from the remote URL rather than the full response being returned.
3. A lot easier to interact with than a `WP_Rest_Response` object.

Reply:
@TimothyBJacobs How else would someone add additional fields to the response while having access to the `$response_body`?

For example, I'd want to add an `image` field which grabs an image from the page's source.

Reply:
Ah, my mistake I didn't notice the http response was being added to the "prepare" filter. IMO we should make that available there after `$url`.

Reply:
@TimothyBJacobs Thanks for your suggestion. I just want to clarify we're on the same lines here:

Initially, I'm not _sure_ (I could be wrong!) what you suggest above will address the issues I raised in https://github.com/WordPress/gutenberg/pull/18042#discussion_r555842355.

The `rest_url_details_url_data` filter allows a developer to _easily_ interrogate the HTTP response from the remote URL and prepare _additional_ items in the `$data` array.

For example:

```php
add_filter(
    'rest_url_details_url_data',
    function($data, $response_body) {
        // parse_og_title is a function I made up for illustration purposes!
        return $data['og_title'] = parse_og_title( $response_body );
    }
);
```

If we add the remote URL HTTP `$response_body` as a argument to the `rest_prepare_url_details` filter like this:

```php
// $response_body is the HTTP response from the remote URL (eg: HTML).
return apply_filters( 'rest_prepare_url_details', $response, $url, $request, $response_body );
```

...then it's going to make it _hard_ for the developer to achieve the same thing because they will have to modify the `WP_REST_Response` object to insert the new `$data` there. Indeed, [I can't see any methods on that class](https://developer.wordpress.org/reference/classes/wp_rest_response/) that allow us to do that.

For example:

```php
add_filter(
    'rest_prepare_url_details',
    function(WP_REST_Response $response, $url, $request, $response_body) {
        $additional_data = array(
            'og_title' => parse_og_title( $response_body )
        );

        // Now I have the additional data how do I append it to the $response?????
    }
);
```

@TimothyBJacobs I have two questions

1. Can you advise on how we could achieve the goal of allow developers to parse additional data out of the remote URL HTTP response to include in the rest response using the `rest_prepare_url_details` filter only (as you propose)?
2. Do you feel the additional complexity (subjective) posed by doing this via `rest_prepare_url_details` is worth it?

I appreciate your time here. I feel like we're nearly there but I do want to get this last piece right because I believe it's an important part of the puzzle in allowing this endpoint to be truly flexible. 🙇

Reply:
Sure!

They can do `$response->data['my_custom_data'] = "value"`;

I'd just switch the order of `$request, $response_body` in the filter.

Reply:
Ah `$response->data`, like I've been doing the whole time in the unit tests! Whoops. 

Right, that's clear now. Let me sort it. Thanks!

Reply:
@TimothyBJacobs As discussed in a DM on Slack, after looking again I've realised the approach suggested will not wor. This is because we do not cache the **full response from the remote URL**. Rather we cache what has been parsed from the response.

When returning the value from _cache_, the `rest_prepare_url_details` filter does not have access to the response from the remote URL. Therefore we cannot use the `rest_prepare_url_details` to parse extra data from the remote URL.

Options:

1. Don't use a cache and pass the remote URL response to the `rest_prepare_url_details` filter.
2. Retain use of a cache and continue to use x2 filters (as currently in the code of this PR).

I'm inclined towards `#2`. Could we go with that in order to get this PR merged. It is after all experimental and can be adjusted in further followups if necessary.

Perhaps you could let me know? Many thanks

Reply:
My concern with the double filtering in the REST controllers is that we're introducing an atypical pattern for modifying a core REST API response. This is further complicated by the fact that the "traditional" filter isn't particularly helpful without the URL response.

IMO it is important that the cache be as transparent to plugin developers as possible. By storing the parsed URL details in the cache, I think we break this transparency. If there is an existing cache hit, a developer won't see their filtered data until the cache expires. This is more problematic if they are utilizing additional request parameters to customize the return value since the parsed URL details may have been cached from a request that didn't have those request parameters.

I think not caching the response is fairly safe since this is a privileged endpoint and is done in response to a fairly explicit user action. However, if we don't want to go that route, IMO we should just cache the raw HTTP response. That would allow for utilizing the single filter that developers would traditionally reach for, and should result in a transparent caching layer.

Reply:
> IMO we should just cache the raw HTTP response

I've updated to implement this approach.

> That would allow for utilizing the single filter that developers would traditionally reach for, and should result in a transparent caching layer.

Done.